hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestAsyncTableScanMetrics.java

```diff
@@ -110,56 +112,101 @@ public static void tearDown() throws Exception {
     UTIL.shutdownMiniCluster();
   }
 
-  private static Pair<List<Result>, ScanMetrics> doScanWithRawAsyncTable(Scan scan)
-    throws IOException, InterruptedException {
+  private static Pair<List<Result>, Pair<ScanMetrics, List<ScanMetrics>>>
+    doScanWithRawAsyncTable(Scan scan) throws IOException, InterruptedException {
     BufferingScanResultConsumer consumer = new BufferingScanResultConsumer();
     CONN.getTable(TABLE_NAME).scan(scan, consumer);
     List<Result> results = new ArrayList<>();
     for (Result result; (result = consumer.take()) != null;) {
       results.add(result);
     }
-    return Pair.newPair(results, consumer.getScanMetrics());
+    return Pair.newPair(results,
+      Pair.newPair(consumer.getScanMetrics(), consumer.getScanMetricsByRegion()));
   }
 
-  private static Pair<List<Result>, ScanMetrics> doScanWithAsyncTableScan(Scan scan)
-    throws Exception {
+  private static Pair<List<Result>, Pair<ScanMetrics, List<ScanMetrics>>>
+    doScanWithAsyncTableScan(Scan scan) throws Exception {
     SimpleScanResultConsumerImpl consumer = new SimpleScanResultConsumerImpl();
     CONN.getTable(TABLE_NAME, ForkJoinPool.commonPool()).scan(scan, consumer);
-    return Pair.newPair(consumer.getAll(), consumer.getScanMetrics());
+    return Pair.newPair(consumer.getAll(),
+      Pair.newPair(consumer.getScanMetrics(), consumer.getScanMetricsByRegion()));
   }
 
-  private static Pair<List<Result>, ScanMetrics> doScanWithAsyncTableScanner(Scan scan)
-    throws IOException {
+  private static Pair<List<Result>, Pair<ScanMetrics, List<ScanMetrics>>>
+    doScanWithAsyncTableScanner(Scan scan) throws IOException {
     try (ResultScanner scanner =
       CONN.getTable(TABLE_NAME, ForkJoinPool.commonPool()).getScanner(scan)) {
       List<Result> results = new ArrayList<>();
       for (Result result; (result = scanner.next()) != null;) {
         results.add(result);
       }
-      return Pair.newPair(results, scanner.getScanMetrics());
+      return Pair.newPair(results,
+        Pair.newPair(scanner.getScanMetrics(), scanner.getScanMetricsByRegion()));
     }
   }
 
   @Test
   public void testNoScanMetrics() throws Exception {
-    Pair<List<Result>, ScanMetrics> pair = method.scan(new Scan());
+    Pair<List<Result>, Pair<ScanMetrics, List<ScanMetrics>>> pair = method.scan(new Scan());
     assertEquals(3, pair.getFirst().size());
-    assertNull(pair.getSecond());
+    // Assert no scan metrics
+    assertNull(pair.getSecond().getFirst());
+    // Assert no per region scan metrics
+    assertNull(pair.getSecond().getSecond());
   }
 
   @Test
   public void testScanMetrics() throws Exception {
-    Pair<List<Result>, ScanMetrics> pair = method.scan(new Scan().setScanMetricsEnabled(true));
```

Comment:
Can you please bust up this method into two variants, one with and one without the per-region accounting enabled? Ideally the latter is a pure super-set of the former, but as things are currently implemented, I think that won't be the case...